Fix currentTopLevelFunction function for metadata skipping

sashton opened the pull request:
<!-- ❤️ Thanks for filing a Pull Request on Calva! You are contributing to a better Clojure coding experience. ❤️
## What you can expect:

Here are some things we consider before we merge:

- We make sure the PR is directed at the `dev` branch (unless reasons).
- We figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and will help you test there if it is hard for you to do so. (We appreciate a lot if you take on the work do this of course.)
- We read the source changes. (Surprise! 😄)
- We given feedback and guidance on source changes, if needed. Far from everything is captured in our [code guidelines](https://github.com/BetterThanTomorrow/calva/wiki/Coding-Style).
- We use our domain knowledge to try catch if you have missed some facility already provided in the code base.
- We read the updates to the documentation and help with feedback, trying to keep the documentation site serving well.
- We often check out your code changes and test them.
- We sometimes send the VSIX built from the PR out in the `#calva` channel on slack for others to test. (Actually, we will probably encourage you to do this.)
- We sometimes have a chat within the team about particular changes.
- NB: We also consider if your changes belong in the Calva product we want to maintain. Before you spend a lot of work on a PR, please consider chatting us up first, and filing issues.

We try to be speedy and attentive. Please don't hesitate to bump a PR, or contact us, if we seem to have dropped the ball (that has happened).

We use checklists in order to not forget about important lessons we and others have learnt along the way.

-->

## What has Changed?
<!-- Introduce the change(s) briefly here. Consider explaining why a particular change was implemented the way it was. If you have considered alternative ways to introduce the change, please elaborate a bit on that as well. -->

~- This PR updates the forwardSexp function for the case where there exists metadata between two `id` nodes and `skipMetadata==true`. Previously, this function would move forward through whitespace and metadata, finally landing on an `id` node. However, in the `switch (token.type) ... case 'id'` code there would be one final call to `this.next()`, thereby moving the cursor past the correct s-expr.~
~- Now, the `switch (token.type) ... case 'id'` scenario has been updated to differentiate between:~
~- the initial iteration of the while loop, where the cursor is still on the original `id` node, and it _is_ correct to move past this original `id` node~
~- any subsequent iteration of the loop, where it is not correct to move the cursor past the current `id` node~

Update:
- The original solution of updating fowardSexp caused unintended consequences with other cursor-motion scenarios. I have reverted my changes there, and instead made the change in `currentTopLevelFunction`. 
- This change updates `currentTopLevelFunction` by checking whether the cursor is on metadata, and, if so, moves forward via `fowardSexp` until a non-metadata node is found. It is a simpler, more specific solution than the first attempt.


<!-- Tell us what Github issue(s) your PR is fixing. Consider creating the issue if there isn't one already. -->

Fixes #1463

## My Calva PR Checklist
<!-- Strike out (using `~`) items that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     ~- [ ] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.~
- [x] Created the issue I am fixing/addressing, if it was not present.

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

Ping @pez, @bpringe

sashton committed:
Fix forwardSexp function for metadata skipping

PEZ commented:
Thanks!

Some tests are failing now. Do you know how to run the tests locally? Let me know otherwise and I'll show you.

sashton commented:
@PEZ I see the failures. I'm investigating. I may need to solve this a different way. From reading the failed tests, I think I had an incorrect mental model of how forwardSexp was designed to operate over metadata and reader tags.

sashton committed:
Revert "Fix forwardSexp function for metadata skipping"

This reverts commit a6dcca117b08b92c996bb9c3847e75ac9d1caf9a.

PEZ commented:
Those parts are quite messy. We should clean it up some day.

sashton committed:
Update currentTopLevelFunction to handle metadata on function symbol

sashton commented:
@PEZ I've reverted the original change and instead made the change in `currentTopLevelFunction`. All tests now pass, including the new one I added to cover the scenario in the bug report.

sashton committed:
Add note to changelog

sashton committed:
Merge branch 'dev' into custom-command-with-fn-metadata

PEZ commented:
I'm having a look and a think about this. You have probably found the right place to do this, but at the same time this function is also very central to a lot of things. And the current tests do pass. I just need to experiment and think a bit. Please don't hesitate to bump this if I go silent for more than a few days.

PEZ commented:
Right so I did go silent... But now have had a look and a think. This seems like the right way to solve the problem so I am going to go ahead and merge. Many thanks for helping in improving Calva!
